### PR TITLE
MetalK8s UI: Implement median percentile chart when the number of node is more than 8

### PR DESCRIPTION
**Pull request by ChengYanJin**
**Component**: UI, monitoring

**Context**: 

**Summary**:
Add feature toggle “force_quantile_chart" in config.json flags:[]  to force displaying quantile chart.

**Acceptance criteria**: 
image

Quantile Chart:
image

---

Closes: ARTESCA-2204

### CODE REVIEW

File: ui/src/components/NonSymmetricalQuantileChart.js [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+  renderQuantileData,
+  renderTooltipSerie,
+} from '../services/graphUtils';
+
+const NonSymmetricalQuantileChart = ({
+  getQuantileQuery,
+  getQuantileHoverQuery,
+  title,
+  yAxisType,
+  isLegendHided,
```
Comment (JBWatenbergScality) [inline ui/src/components/NonSymmetricalQuantileChart.js]: we should likely rename this to isLegendHidden

File: ui/src/components/NonSymmetricalQuantileChart.js [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+      ([
+        prometheusResultQuantile5,
+        prometheusResultMedian,
+        prometheusResultQuantile90,
+      ]) => {
+        return [
+          convertPrometheusResultToSerie(prometheusResultQuantile90, 'Q90'),
+          convertPrometheusResultToSerie(prometheusResultMedian, 'Median'),
+          convertPrometheusResultToSerie(prometheusResultQuantile5, 'Q5'),
+        ];
```
Comment (JBWatenbergScality) [inline ui/src/components/NonSymmetricalQuantileChart.js]: nit: Just to sort things a little bit, I would reorder the query array to match the "rendering" order

File: ui/src/components/NonSymmetricalQuantileChart.js [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+      isLoading={isLoadingQuantile}
+      onHover={onHover}
+      renderTooltipSerie={useCallback(
+        (serie, tooltipData) => {
+          if (serie.key === 'Q90') {
+            return (
+              renderTooltipSerie(serie) +
+              `<tr style="color: ${
+                theme.textSecondary
+              }"><td></td><td colspan="2" style="padding-left: 1rem;">Nodes above Q90</td></tr>
```
Comment (JBWatenbergScality) [inline ui/src/components/NonSymmetricalQuantileChart.js]: Maybe we should replace Q90 per `serie.key` and then I wonder if the condition can be collapsed with the one below.

File: ui/src/components/SymmetricalQuantileChart.js [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+                isLoadingQuantile5Out,
+                isSuccessQuantile5Out,
+                quantile5OutData,
+                nodeMapPerIp,
+                theme,
+                valueBase,
+                serie.unitLabel,
+              )}`
+            );
+          }
```
Comment (JBWatenbergScality) [inline ui/src/components/SymmetricalQuantileChart.js]: I wonder if we should/could collapse those conditions because in the end just a few params/words are changing

File: ui/src/components/DashboardBandwidthChart.js [hunk 5]
```diff
@@ -42,18 +47,20 @@ const DashboardBandwidthChart = ({
...
-    getQueryAbove: (timeSpanProps) =>
+    getAboveQueries: (timeSpanProps) => [
       getNodesPlanesBandwidthInQuery(timeSpanProps, devices),
-    getQueryBelow: (timeSpanProps) =>
+    ],
+    getBelowQueries: (timeSpanProps) => [
       getNodesPlanesBandwidthOutQuery(timeSpanProps, devices),
+    ],
     transformPrometheusDataToSeries: useCallback(
       (prometheusResultAbove, prometheusResultBelow) => {
```
Comment (JBWatenbergScality) [inline ui/src/components/DashboardBandwidthChart.js]: Instead of adding a lot of [0] bellow I would just change this like below:
```
([prometheusResultAbove], [prometheusResultBelow]) => {
```

File: ui/src/components/DashboardNetwork.js [hunk 6]
```diff
@@ -29,24 +31,32 @@ const ChartContainer = styled.div`
 
 const DashboardNetwork = () => {
   const intl = useIntl();
-
+  const nodes = useNodes();
+  const { flags } = useTypedSelector((state) => state.config.api);
+  const isShowQuantileChart =
+    (flags && flags.includes('force_quantile_chart')) ||
+    nodes?.length > NODES_LIMIT_QUANTILE;
```
Comment (JBWatenbergScality) [inline ui/src/components/DashboardNetwork.js]: I would create a custom hook for that to mutualise the logic with `ui/src/components/DashboardMetrics.js`. Generally speaking I would tend to avoid as much as possible duplicating the same logic in order to ease evolution (DRY - Don't Repeat Yourself - principle for the win 😉 )

File: ui/public/config.json [hunk 7]
```diff
@@ -5,6 +5,7 @@
   "url_grafana": "/grafana",
   "url_doc": "/docs",
   "url_alertmanager": "/api/alertmanager",
+  "url_loki": "/api/loki",
```
Comment (JBWatenbergScality) [inline ui/public/config.json:8]: nit: The commit message is mentioning  force_quantile_chart flag but I don't see it here

File: ui/src/services/platformlibrary/metrics.js [hunk 8]
```diff
@@ -82,6 +82,61 @@ export const getNodesCPUUsageQuery = (timespanProps: TimeSpanProps) => {
...
+        });
+      }
+      return Promise.reject();
+    },
+    refetchOnMount: false,
+    refetchOnWindowFocus: false,
+    enabled: !!(timestamp && threshold !== undefined),
+  };
+};
+
```
Comment (JBWatenbergScality) [inline ui/src/services/platformlibrary/metrics.js]: I would factorise QuantileQuery and threshold queries implementation using a function private to this module to avoid repeating the same prometheus result parsing over and over.

Comment (JBWatenbergScality) [reply ui/src/services/platformlibrary/metrics.js]: Something like: 
```js
export const getNodesCPUUsageQuantileQuery = (
  timespanProps: TimeSpanProps,
  quantile: number,
) => {
  return _getQuantileQuery({getPromQuery: ({startingTimeISO, currentTimeISO, frequency}) => `quantile(${quantile}, 100 - (avg by (instance) (irate(node_cpu_seconds_total{mode="idle"}[5m])) * 100))`, getQueryKey: ({startingTimeISO, currentTimeISO, frequency}) => ['NodesCpuUsageQuantile', startingTimeISO, quantile]})
};

export const getNodesCPUUsageAboveBelowThresholdQuery = (
  timestamp?: string,
  threshold?: number,
  operator: '>' | '<',
) => {
  return _getOutpassingThresholdQuery({getPromQuery: ({startingTimeISO, currentTimeISO, frequency}) => `100 - (avg by (instance) (irate(node_cpu_seconds_total{mode="idle"}[5m])) * 100) ${operator}= ${threshold}`}, getQueryKey: ({startingTimeISO, currentTimeISO, frequency}) => [
      'NodesCpuUsageAboveBelowThreshold',
      timestamp,
      threshold,
      operator,
    ]});
};
```

File: ui/src/components/VolumeCharts.js [hunk 9]
```diff
@@ -30,15 +30,17 @@ export const VolumeThroughputChart = ({
...
-    getQueryAbove: (timeSpanProps: TimeSpanProps) =>
+    getAboveQueries: (timeSpanProps: TimeSpanProps) => [
       getVolumeThroughputWriteQuery(instanceIp, deviceName, timeSpanProps),
-    getQueryBelow: (timeSpanProps: TimeSpanProps) =>
+    ],
+    getBelowQueries: (timeSpanProps: TimeSpanProps) => [
       getVolumeThroughputReadQuery(instanceIp, deviceName, timeSpanProps),
+    ],
     transformPrometheusDataToSeries: useCallback(
       (prometheusResultAbove, prometheusResultBelow) =>
```
Comment (JBWatenbergScality) [inline ui/src/components/VolumeCharts.js]: `([prometheusResultAbove], [prometheusResultBelow]) =>` that would avoid `[0]` accessors below

File: ui/src/components/VolumeCharts.js [hunk 10]
```diff
@@ -72,15 +74,17 @@ export const VolumeLatencyChart = ({
...
-    getQueryAbove: (timeSpanProps: TimeSpanProps) =>
+    getAboveQueries: (timeSpanProps: TimeSpanProps) => [
       getVolumeLatencyWriteQuery(instanceIp, deviceName, timeSpanProps),
-    getQueryBelow: (timeSpanProps: TimeSpanProps) =>
+    ],
+    getBelowQueries: (timeSpanProps: TimeSpanProps) => [
       getVolumeLatencyReadQuery(instanceIp, deviceName, timeSpanProps),
+    ],
     transformPrometheusDataToSeries: useCallback(
       (prometheusResultAbove, prometheusResultBelow) =>
```
Comment (JBWatenbergScality) [inline ui/src/components/VolumeCharts.js]: `([prometheusResultAbove], [prometheusResultBelow]) =>`

File: ui/src/services/graphUtils.js [hunk 11]
```diff
@@ -205,3 +302,80 @@ export const getNodesInterfacesString = (nodeIPsInfo): [] => {
...
+  };
+
+  return `${
+    isIdle || isLoading
+      ? `<tr style="color: ${theme.textSecondary}"><td></td><td colspan='2' style="padding-left: 1rem;">Loading...</td></tr>`
+      : ''
+  }
+  ${
+    isSuccess
+      ? data?.data?.result
```
Comment (JBWatenbergScality) [inline ui/src/services/graphUtils.js]: I think we should display an inline message to the user if we fail to load the data for some reason

**Review by JBWatenbergScality — CHANGES_REQUESTED** [review]

File: ui/src/services/platformlibrary/metrics.js (hunk 8 above)
Comment (ChengYanJin) [reply ui/src/services/platformlibrary/metrics.js]: Added `_getPromRangeMatrixQuery` and `_getInstantValueQuery` in a945bfff53da93e5df72c37eec728ad857dd5a83

File: ui/src/services/graphUtils.js (hunk 11 above)
Comment (ChengYanJin) [reply ui/src/services/graphUtils.js]: Add `Sorry, failed to load data due to error occurs.` when isSuccess is false

File: ui/src/components/NonSymmetricalQuantileChart.js (hunk 1 above)
Comment (ChengYanJin) [reply ui/src/components/NonSymmetricalQuantileChart.js]: Rename this prop in Linetemporalchart in core-ui

File: ui/src/components/DashboardBandwidthChart.js [hunk 12]
```diff
@@ -76,6 +80,18 @@ const DashboardBandwidthChartWithoutQuantile = ({
...
+        (serie) => {
+          if (serie.key === `${lastNodeName}-in`) {
+            // add a seperation line between in and out
+            return `${defaultRenderTooltipSerie(serie)}</table>
+            <hr style="border-color: ${theme.border};"/><table>`;
+          } else {
+            return defaultRenderTooltipSerie(serie);
+          }
+        },
+        [lastNodeName, theme],
```
Comment (JBWatenbergScality) [inline ui/src/components/DashboardBandwidthChart.js]: This seems to be duplicated in ui/src/components/DashboardChartThroughput.js

Comment (JBWatenbergScality) [reply ui/src/components/DashboardBandwidthChart.js]: and partly here ui/src/components/SymmetricalQuantileChart.js

File: ui/src/services/graphUtils.js [hunk 13]
```diff
@@ -392,6 +392,6 @@ export const renderQuantileData = (
               )}</td></tr>`,
           )
           .join('')
-      : ''
+      : 'Sorry, failed to load data due to error occurs.'
```
Comment (JBWatenbergScality) [inline ui/src/services/graphUtils.js]: It shall be translated. Also I would reword it to : 
EN : `Sorry an error occurred and we failed to retrieve the nodes outpassing this threshold.`
FR : `Désolé une erreur nous empêche de retrouver les noeuds dépassant ce seuil.`

File: ui/src/components/DashboardBandwidthChart.js (hunk 12 above)
Comment (ChengYanJin) [reply ui/src/components/DashboardBandwidthChart.js]: Extract renderTooltipSeperationLine() in 43ed23753047abf4996610526fa3a7faed579a39

File: ui/src/services/graphUtils.js (hunk 13 above)
Comment (ChengYanJin) [reply ui/src/services/graphUtils.js]: Add the translation in 92f6b474ebcf1c6fe5a7bddb86186991c3278ce6

**Review by JBWatenbergScality — APPROVED** [review]
